pkg/auth/handler/login_auth_url.go

```diff
@@ -0,0 +1,136 @@
+package handler
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/sso"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+
+	"github.com/skygeario/skygear-server/pkg/auth"
+	coreAuth "github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz/policy"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	"github.com/skygeario/skygear-server/pkg/core/inject"
+	"github.com/skygeario/skygear-server/pkg/core/server"
+)
+
+func AttachLoginAuthURLHandler(
+	server *server.Server,
+	authDependency auth.DependencyMap,
+) *server.Server {
+	server.Handle("/sso/{provider}/login_auth_url", &LoginAuthURLHandlerFactory{
+		authDependency,
+	}).Methods("OPTIONS", "POST")
+	return server
+}
+
+type LoginAuthURLHandlerFactory struct {
+	Dependency auth.DependencyMap
+}
+
+func (f LoginAuthURLHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &LoginAuthURLHandler{}
+	inject.DefaultInject(h, f.Dependency, request)
+	vars := mux.Vars(request)
+	h.ProviderName = vars["provider"]
+	return handler.APIHandlerToHandler(h, h.TxContext)
+}
+
+func (f LoginAuthURLHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return authz.PolicyFunc(policy.DenyNoAccessKey)
+}
+
+// LoginAuthURLRequestPayload login handler request payload
+type LoginAuthURLRequestPayload struct {
+	Scope       []string               `json:"scope"`
+	Options     map[string]interface{} `json:"options"`
+	CallbackURL string                 `json:"callback_url"`
+	RawUXMode   string                 `json:"ux_mode"`
+	UXMode      sso.UXMode
+}
+
+// Validate request payload
+func (p LoginAuthURLRequestPayload) Validate() error {
+	if p.CallbackURL == "" {
+		return skyerr.NewInvalidArgument("Callback url is required", []string{"callback_url"})
+	}
+
+	if p.UXMode == sso.Undefined {
+		return skyerr.NewInvalidArgument("UX mode is required", []string{"ux_mode"})
+	}
+
+	return nil
+}
+
+// LoginAuthURLHandler returns the SSO auth url by provider.
+//
+// curl \
+//   -X POST \
+//   -H "Content-Type: application/json" \
+//   -H "X-Skygear-Api-Key: API_KEY" \
+//   -d @- \
+//   http://localhost:3000/sso/<provider>/login_auth_url \
+// <<EOF
+// {
+//     "scope": ["openid", "profile"],
+//     "options": {
+//       "prompt": "select_account"
+//     },
+//     callback_url: <url>,
+//     ux_mode: <ux_mode>
+// }
+// EOF
+//
+// {
+//     "result": "<auth_url>"
+// }
+type LoginAuthURLHandler struct {
+	TxContext    db.TxContext           `dependency:"TxContext"`
+	AuthContext  coreAuth.ContextGetter `dependency:"AuthContextGetter"`
+	Provider     sso.Provider           `dependency:"SSOProvider,optional"`
```

Comment:
`Provider` is not actually `optional` here, because throwing error is not a normal flow, otherwise any dependency can be optional and the handlers would be polluted with error handling.

I think the error for unrecognised provider should be thrown at `inject.go`.

Reply:
@Steven-Chan Updated. Thanks.